### PR TITLE
Qt: Implement advanced mapping I/O windows

### PR DESCRIPTION
**Pull request by spycrab**
Screenshots:

### CODE REVIEW

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+  m_button_box->addButton(m_clear_button, QDialogButtonBox::ActionRole);
+  m_button_box->addButton(m_apply_button, QDialogButtonBox::ActionRole);
+  m_button_box->addButton(QDialogButtonBox::Ok);
+
+  setLayout(m_main_layout);
+}
+
+void IOWindow::Update()
+{
+  if (m_controller == nullptr)
```
Comment (ligfx) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp]: How does this work? Isn't the parent window disabled while this window is visible?

File: Source/Core/DolphinQt2/Config/Mapping/MappingButton.cpp [hunk 2]
```diff
@@ -2,24 +2,38 @@
...
+#include "DolphinQt2/Config/Mapping/MappingCommon.h"
 #include "DolphinQt2/Config/Mapping/MappingWidget.h"
 #include "DolphinQt2/Config/Mapping/MappingWindow.h"
 #include "InputCommon/ControlReference/ControlReference.h"
 #include "InputCommon/ControllerEmu/ControllerEmu.h"
 #include "InputCommon/ControllerInterface/ControllerInterface.h"
+#include "InputCommon/ControllerInterface/Device.h"
+
+static QString TrunicateDisplayString(const QString& string)
+{
```
Comment (spycrab) [inline Source/Core/DolphinQt2/Config/Mapping/MappingButton.cpp]: Sorry but that seems very overkill.
Also I think it kinda belongs into this PR as it allows (the mythical) Qt User to actually enter such long combinations. I think it just barely fits the scope.

File: Source/Core/DolphinQt2/Config/Mapping/MappingButton.cpp [hunk 3]
```diff
@@ -2,24 +2,38 @@
...
+    return string;
+  else
+    return string.left(13).append(QStringLiteral("..."));
+}
 
 MappingButton::MappingButton(MappingWidget* widget, ControlReference* ref)
-    : QPushButton(QString::fromStdString(ref->expression)), m_parent(widget), m_reference(ref)
+    : QPushButton(TrunicateDisplayString(QString::fromStdString(ref->expression))),
+      m_parent(widget), m_reference(ref),
+      m_io_window(
```
Comment (spycrab) [inline Source/Core/DolphinQt2/Config/Mapping/MappingButton.cpp]: I have the feeling it would be just more convoluted with little to no benefit.

File: Source/Core/DolphinQt2/Config/Mapping/MappingButton.cpp [hunk 4]
```diff
@@ -138,25 +124,38 @@ bool MappingButton::event(QEvent* event)
...
+      m_io_window->Update();
+      m_io_window->exec();
+      Update();
+    }
+    return;
+  case Qt::MouseButton::MiddleButton:
+    Clear();
+    break;
+  case Qt::MouseButton::RightButton:
+    if (m_parent->GetParent()->GetDevice() == nullptr)
```
Comment (spycrab) [inline Source/Core/DolphinQt2/Config/Mapping/MappingButton.cpp]: Is this related to the suggestion above? If it is, yes that would only be logical then.

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp (hunk 1 above)
Comment (spycrab) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp]: Can you rephrase that please? I don't really get what you are trying to say.

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+#include <QHBoxLayout>
+#include <QLabel>
+#include <QListWidget>
+#include <QPlainTextEdit>
+#include <QPushButton>
+#include <QSlider>
+#include <QSpinBox>
+#include <QVBoxLayout>
+
+static constexpr int SLIDER_TICK_COUNT = 100;
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp]: `static` isn't necessary with file-scoped constants (const internally links by default).

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+  m_range_slider->setMinimum(-500);
+  m_range_slider->setMaximum(500);
+  m_range_spinbox->setMinimum(-500);
+  m_range_spinbox->setMaximum(500);
+  m_main_layout->addItem(range_hbox);
+
+  // Options (Buttons, Outputs) and action buttons
+  for (QPushButton* button : {m_select_button, m_detect_button, m_or_button, m_and_button,
+                              m_add_button, m_not_button, m_test_button})
+    button->setSizePolicy(QSizePolicy::Fixed, QSizePolicy::Expanding);
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp]: This should be braced since the header spans more than one line.

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+{
+  if (m_block)
+    return;
+
+  m_option_list->clear();
+
+  const auto device = g_controller_interface.FindDevice(m_devq);
+  if (m_reference->IsInput())
+  {
+    for (auto* input : device->Inputs())
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp]: `const auto*`

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+    {
+      m_option_list->addItem(QString::fromStdString(input->GetName()));
+    }
+  }
+  else
+  {
+    for (auto* output : device->Outputs())
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp]: ditto

File: Source/Core/DolphinQt2/Config/Mapping/MappingButton.cpp [hunk 9]
```diff
@@ -2,24 +2,38 @@
...
+#include "DolphinQt2/Config/Mapping/IOWindow.h"
+#include "DolphinQt2/Config/Mapping/MappingCommon.h"
 #include "DolphinQt2/Config/Mapping/MappingWidget.h"
 #include "DolphinQt2/Config/Mapping/MappingWindow.h"
 #include "InputCommon/ControlReference/ControlReference.h"
 #include "InputCommon/ControllerEmu/ControllerEmu.h"
 #include "InputCommon/ControllerInterface/ControllerInterface.h"
+#include "InputCommon/ControllerInterface/Device.h"
+
+static QString TrunicateDisplayString(const QString& string)
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Config/Mapping/MappingButton.cpp]: Truncate

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+#include <QComboBox>
+#include <QDialogButtonBox>
+#include <QHBoxLayout>
+#include <QLabel>
+#include <QListWidget>
+#include <QPlainTextEdit>
+#include <QPushButton>
+#include <QSlider>
+#include <QSpinBox>
+#include <QVBoxLayout>
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp]: The ordering of these includes should be like so:

```cpp
#include "DolphinQt2/Config/Mapping/IOWindow.h"

#include <thread>

#include <QComboBox>
#include <QDialogButtonBox>
#include <QHBoxLayout>
#include <QLabel>
#include <QListWidget>
#include <QPlainTextEdit>
#include <QPushButton>
#include <QSlider>
#include <QSpinBox>
#include <QVBoxLayout>

#include "Core/Core.h"
#include "DolphinQt2/Config/Mapping/MappingButton.h"
#include "DolphinQt2/Config/Mapping/MappingCommon.h"
#include "InputCommon/ControlReference/ControlReference.h"
#include "InputCommon/ControllerEmu/ControllerEmu.h"
#include "InputCommon/ControllerInterface/ControllerInterface.h"
```

File: Source/Core/DolphinQt2/Config/Mapping/MappingCommon.cpp [hunk 11]
```diff
@@ -0,0 +1,34 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "DolphinQt2/Config/Mapping/MappingCommon.h"
+#include "InputCommon/ControllerInterface/Device.h"
+
+#include <QRegExp>
+#include <QString>
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Config/Mapping/MappingCommon.cpp]: The ordering of these includes should be like so:

```cpp
#include "DolphinQt2/Config/Mapping/MappingCommon.h"

#include <QRegExp>
#include <QString>

#include "InputCommon/ControllerInterface/Device.h"
```

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp (hunk 1 above)
Comment (ligfx) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp]: Er, let me read through this some more.

File: Source/Core/DolphinQt2/Config/Mapping/MappingButton.cpp (hunk 3 above)
Comment (ligfx) [inline Source/Core/DolphinQt2/Config/Mapping/MappingButton.cpp]: Would you mind trying it out? The point of Qt's signals-and-slots things is to make it easier to do this sort of thing. If you think about it in the MVC paradigm, you can think of it like a PushButton is a dumb view, so you don't want it owning other views, and want the controller/model to handle all of the logic.

File: Source/Core/DolphinQt2/Config/Mapping/MappingButton.cpp (hunk 2 above)
Comment (spycrab) [inline Source/Core/DolphinQt2/Config/Mapping/MappingButton.cpp]: Hm, it might be worth the time investment if it were used elsewhere, but I don't think it really is.

File: Source/Core/DolphinQt2/Config/Mapping/MappingButton.cpp (hunk 3 above)
Comment (spycrab) [inline Source/Core/DolphinQt2/Config/Mapping/MappingButton.cpp]: Will try it.

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp (hunk 10 above)
Comment (spycrab) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp]: Reordered. Although GitHub does not think so.

**Comment by spycrab** [comment]
@ligfx Can you take a look again?

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+
+void IOWindow::UpdateOptionList()
+{
+  if (m_block)
+    return;
+
+  m_option_list->clear();
+
+  const auto device = g_controller_interface.FindDevice(m_devq);
+  if (m_reference->IsInput())
```
Comment (ligfx) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp]: If you wanted, you could deduplicate this with something like:

```c++
Control* control_iter = m_reference->IsInput() ? &device->Inputs().front() : &device->Outputs().front();
const Control* control_end = (m_reference->IsInput() ? &device->Inputs().back() : &device->Outputs().back()) + 1;
for (; control_iter != control_end; control_iter++)
{
  m_option_list->addItem(QString::fromStdString(control_iter->GetName()));
}
```

since inputs and outputs are both subclasses of `Device::Control`.

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+
+  const bool paused = Core::PauseAndLock(true);
+
+  g_controller_interface.RefreshDevices();
+  m_controller->UpdateReferences(g_controller_interface);
+  m_controller->UpdateDefaultDevice();
+
+  const auto default_device = m_controller->default_device.ToString();
+
+  m_devices_combo->addItem(QString::fromStdString(default_device));
```
Comment (ligfx) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp]: Can you put a comment on all of this to explain why the default device is added manually?

File: Source/Core/DolphinQt2/Config/Mapping/MappingButton.cpp [hunk 14]
```diff
@@ -2,24 +2,36 @@
...
+#include "DolphinQt2/Config/Mapping/IOWindow.h"
+#include "DolphinQt2/Config/Mapping/MappingCommon.h"
 #include "DolphinQt2/Config/Mapping/MappingWidget.h"
 #include "DolphinQt2/Config/Mapping/MappingWindow.h"
 #include "InputCommon/ControlReference/ControlReference.h"
 #include "InputCommon/ControllerEmu/ControllerEmu.h"
 #include "InputCommon/ControllerInterface/ControllerInterface.h"
+#include "InputCommon/ControllerInterface/Device.h"
+
+static QString TruncateDisplayString(const QString& string)
```
Comment (ligfx) [inline Source/Core/DolphinQt2/Config/Mapping/MappingButton.cpp]: Here's an example of the issue with eliding based on number of characters rather than on available width:

image

If you want, I can go figure out the ElidedButton stuff and put up a separate PR, since this is already a problem with current controller mapping window (in that it doesn't elide at all, and my mappings are giant-ass strings like `Up Arrow + Input/0/Xbox One Wired Controller:Axis Y-`).

Comment (spycrab) [inline Source/Core/DolphinQt2/Config/Mapping/MappingButton.cpp]: Yeah, it's probably better to do that in a separate PR.

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+    accept();
+}
+
+void IOWindow::OnDetectButtonPressed()
+{
+  if (m_block)
+    return;
+
+  m_block = true;
+  std::thread([this] {
```
Comment (spycrab) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp]: Shared code now moved to ``MappingCommon``

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp (hunk 12 above)
Comment (spycrab) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp]: Would probably be better, but I think the current version is a lot more readable and just as efficient.

Comment (ligfx) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp]: Fair enough, too bad you can't easily upcast vectors :/

File: Source/Core/DolphinQt2/Config/Mapping/MappingWidget.cpp [hunk 16]
```diff
@@ -49,10 +50,22 @@ QGroupBox* MappingWidget::CreateGroupBox(const QString& name, ControllerEmu::Con
...
+    auto* control_ref = control->control_ref.get();
+
+    connect(button, &MappingButton::AdvancedPressed, [this, &iowindow, control_ref] {
+      iowindow = new IOWindow(this, control_ref, control_ref->IsInput() ? IOWindow::Type::Input :
+                                                                          IOWindow::Type::Output);
+      iowindow->Open();
+    });
+
     m_buttons.push_back(button);
+    m_iowindows.push_back(iowindow);
```
Comment (ligfx) [inline Source/Core/DolphinQt2/Config/Mapping/MappingWidget.cpp]: So this is a vector of a bunch of `nullptr`s?

File: Source/Core/DolphinQt2/Config/Mapping/MappingWidget.cpp [hunk 17]
```diff
@@ -49,10 +50,22 @@ QGroupBox* MappingWidget::CreateGroupBox(const QString& name, ControllerEmu::Con
...
     button->setMinimumWidth(125);
     button->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Fixed);
     form_layout->addRow(QString::fromStdString(control->name), button);
+
+    auto* control_ref = control->control_ref.get();
+
+    connect(button, &MappingButton::AdvancedPressed, [this, &iowindow, control_ref] {
+      iowindow = new IOWindow(this, control_ref, control_ref->IsInput() ? IOWindow::Type::Input :
+                                                                          IOWindow::Type::Output);
+      iowindow->Open();
```
Comment (ligfx) [inline Source/Core/DolphinQt2/Config/Mapping/MappingWidget.cpp]: `iowindow` should be deleted, since it's not stored anywhere here.

File: Source/Core/DolphinQt2/Config/Mapping/MappingWidget.cpp [hunk 18]
```diff
@@ -87,6 +100,12 @@ void MappingWidget::OnClearFields()
 
 void MappingWidget::Update()
 {
+  for (auto* io : m_iowindows)
```
Comment (ligfx) [inline Source/Core/DolphinQt2/Config/Mapping/MappingWidget.cpp]: These are all `nullptr`s, so this is unnecessary.

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp [hunk 19]
```diff
@@ -0,0 +1,269 @@
...
+  // Button Box
+  m_main_layout->addWidget(m_button_box);
+  m_button_box->addButton(m_clear_button, QDialogButtonBox::ActionRole);
+  m_button_box->addButton(m_apply_button, QDialogButtonBox::ActionRole);
+  m_button_box->addButton(QDialogButtonBox::Ok);
+
+  setLayout(m_main_layout);
+}
+
+void IOWindow::Update()
```
Comment (ligfx) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp]: Since this is only used in `Open()`, it can be inlined there.

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp [hunk 20]
```diff
@@ -0,0 +1,269 @@
...
+#include "DolphinQt2/Config/Mapping/MappingWidget.h"
+#include "DolphinQt2/Config/Mapping/MappingWindow.h"
+#include "InputCommon/ControlReference/ControlReference.h"
+#include "InputCommon/ControllerEmu/ControllerEmu.h"
+#include "InputCommon/ControllerInterface/ControllerInterface.h"
+
+constexpr int SLIDER_TICK_COUNT = 100;
+
+IOWindow::IOWindow(MappingWidget* parent, ControlReference* ref, IOWindow::Type type)
+    : QDialog(static_cast<QWidget*>(parent)), m_reference(ref), m_parent(parent), m_type(type)
```
Comment (ligfx) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp]: Since this blocks the parent window while in use and is never reused, there's no need to call functions/get data from the parent. Any needed data should be passed in explicitly.

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp [hunk 21]
```diff
@@ -0,0 +1,269 @@
...
+}
+
+void IOWindow::Open()
+{
+  if (m_parent->GetParent()->GetDevice() == nullptr)
+    return;
+
+  Update();
+  exec();
+  m_parent->Update();
```
Comment (ligfx) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp]: This line should be moved to `MappingWidget`.

File: Source/Core/DolphinQt2/Config/Mapping/MappingCommon.cpp [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+namespace MappingCommon
+{
+QString GetExpressionForControl(const QString& control_name,
+                                const ciface::Core::DeviceQualifier& control_device,
+                                const ciface::Core::DeviceQualifier& default_device)
+{
+  QString expr;
+
+  // non-default device
+  // operator== is overloaded, but operator!= isn't
```
Comment (ligfx) [inline Source/Core/DolphinQt2/Config/Mapping/MappingCommon.cpp]: What's preventing `operator!=` from being overloaded?

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp (hunk 19 above)
Comment (lioncash) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp]: This should likely be kept separate, since it doesn't directly relate to the opening of the dialog.

File: Source/Core/DolphinQt2/Config/Mapping/MappingCommon.cpp (hunk 22 above)
Comment (spycrab) [inline Source/Core/DolphinQt2/Config/Mapping/MappingCommon.cpp]: Nothing that I know of, but it currently isn't.

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp (hunk 19 above)
Comment (ligfx) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp]: Hmm, you're right, good point. This can be made `private` at least though.

**Comment by spycrab** [comment]
Would you mind taking a look again, @ligfx?

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.h [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+  // Buttonbox
+  QDialogButtonBox* m_button_box;
+  QPushButton* m_clear_button;
+  QPushButton* m_apply_button;
+
+  ControlReference* m_reference;
+  ControllerEmu::EmulatedController* m_controller;
+
+  ciface::Core::DeviceQualifier m_devq;
+  bool m_block = false;
```
Comment (ligfx) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.h]: Should be `std::atomic`

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp [hunk 24]
```diff
@@ -0,0 +1,259 @@
...
+#include "DolphinQt2/Config/Mapping/MappingCommon.h"
+#include "DolphinQt2/Config/Mapping/MappingWidget.h"
+#include "DolphinQt2/Config/Mapping/MappingWindow.h"
+#include "InputCommon/ControlReference/ControlReference.h"
+#include "InputCommon/ControllerEmu/ControllerEmu.h"
+#include "InputCommon/ControllerInterface/ControllerInterface.h"
+
+constexpr int SLIDER_TICK_COUNT = 100;
+
+IOWindow::IOWindow(MappingWidget* parent, ControllerEmu::EmulatedController* controller,
```
Comment (ligfx) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp]: `parent` can be a `QWidget*`, removing the need for the cast below

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.h [hunk 25]
```diff
@@ -0,0 +1,98 @@
...
+{
+  Q_OBJECT
+public:
+  enum class Type
+  {
+    Input,
+    Output
+  };
+
+  explicit IOWindow(MappingWidget* parent, ControllerEmu::EmulatedController* m_controller,
```
Comment (ligfx) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.h]: Does this need to be `explicit`?

Comment (spycrab) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.h]: I'm not sure.

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp [hunk 26]
```diff
@@ -0,0 +1,259 @@
...
+
+    if (!expr.isEmpty())
+    {
+      const auto list = m_option_list->findItems(expr, Qt::MatchFixedString);
+
+      if (list.size() > 0)
+        m_option_list->setCurrentItem(list[0]);
+    }
+    m_block = false;
+  }).detach();
```
Comment (shuffle2) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp]: detach doesn't seem like a good idea?

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.h (hunk 25 above)
Comment (ligfx) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.h]: I'm sure @lioncash knows!

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp (hunk 26 above)
Comment (spycrab) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp]: "Separates the thread of execution from the thread object, allowing execution to continue independently. Any allocated resources will be freed once the thread exits."

But I'm far from a thread expert, so correct me if I'm wrong, but looks okay to me.

Comment (ligfx) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp]: It is correct, though not a good idea. This is how the existing `MappingButton` code was, though, so might be better to fix it all in one fell swoop in a separate PR.

**Comment by MayImilae** [comment]
Well, I tried to test this to make sure everything is working, but Qt (at least in the build I just downloaded, the builder is having trouble) doesn't have any indicator lights of when buttons are pressed! Without that ability, testing would require running a game and seeing if it responds, and that's MUCH harder!

I think it might be a good idea to hold off on merging this until after button press notification is added, then it will be a LOT easier to test this PR and make sure everything is working correctly.

File: Source/Core/DolphinQt2/Config/Mapping/IOWindow.h (hunk 25 above)
Comment (lioncash) [inline Source/Core/DolphinQt2/Config/Mapping/IOWindow.h]: In the case of multiple arguments, it depends. For example, if you want to prevent things like so from occurring: [examples](https://ideone.com/7yN37Q) then it should likely be explicit. Constructors without `explicit` on them are considered [converting constructors](http://en.cppreference.com/w/cpp/language/converting_constructor)

**Comment by spycrab** [comment]
@MayImilae I disagree: The only thing worth testing is whether it's creating correct expressions and I think most cases can be handled with one single expression that uses everything once.

**Comment by MayImilae** [comment]
Well, as long as you know it works!

**Comment by spycrab** [comment]
@lioncash Any other suggestions?

File: Source/Core/DolphinQt2/Config/Mapping/MappingButton.cpp [hunk 27]
```diff
@@ -137,25 +112,21 @@ bool MappingButton::event(QEvent* event)
...
-  }
-  else
-  {
-    // TODO Open output dialog
+    else
+      emit AdvancedPressed();
+    return;
+  case Qt::MouseButton::MiddleButton:
+    Clear();
+    break;
```
Comment (leoetlino) [inline Source/Core/DolphinQt2/Config/Mapping/MappingButton.cpp]: for consistency, return?

Comment (spycrab) [inline Source/Core/DolphinQt2/Config/Mapping/MappingButton.cpp]: Sure.
